Lib/test/test_concurrent_futures.py

```diff
@@ -342,6 +342,29 @@ def test_hang_issue12364(self):
         for f in fs:
             f.result()
 
+    def test_cancel_futures(self):
+        executor = self.executor_type()
+        fs = [executor.submit(time.sleep, .1) for _ in range(50)]
+        executor.shutdown(cancel_futures=True)
+        # We can't guarantee the exact number of cancellations, but we can
+        # guarantee that *some* were cancelled. In this case, at least half
+        # should have been cancelled.
```

Comment:
I'm not sure I understand where that guarantee comes from. You could be executing the tests on a very wide machine (for example a 32-thread CPU), and/or thread scheduling details could make the cancel() calls be executed very late.

If you limited the number of workers to a small value in the `executor_type()` constructor call, the test reliability would probably be improved.

Reply:
I had previously specified `max_workers` in this test, but I think that was lost after making a recent round of changes (to improve the tests and fix a likely race condition). I was using 5 before, do you think that's a reasonable baseline? Several of the other tests in `test_concurrent_futures` seem to use that amount.

Reply:
Yes, that's reasonable, though in this guess you could go as low as 3.

Reply:
I set `max_workers=3`, and modified the number of expected cancelled futures to 35 or more (when I added temporary logging, the average seemed to be ~45). This should give plenty of breathing room for different platforms, while still ensuring that `cancel_futures` is functioning as intended.

I also removed the "half" or more guarantee in the comment, as it was fairly arbitrary. My intention was ensure that most of the cancelled futures were cancelled,

I tested it using 100 parallel workers to try accounting for race conditions:
```bash
./python -m test test_concurrent_futures --match test_cancel_futures -j100 -v -F
...
Ran 4 tests in 3.066s

OK
0:01:34 load avg: 114.32 [1033] test_concurrent_futures passed
test_cancel_futures (test.test_concurrent_futures.ProcessPoolForkProcessPoolShutdownTest) ... 0.42s ok
test_cancel_futures (test.test_concurrent_futures.ProcessPoolForkserverProcessPoolShutdownTest) ... 0.52s ok
test_cancel_futures (test.test_concurrent_futures.ProcessPoolSpawnProcessPoolShutdownTest) ... 0.54s ok
test_cancel_futures (test.test_concurrent_futures.ThreadPoolShutdownTest) ... 0.10s ok
...
== Tests result: INTERRUPTED ==
Test suite interrupted by signal SIGINT.

1033 tests OK.

Total duration: 1 min 34 sec
Tests result: INTERRUPTED
```